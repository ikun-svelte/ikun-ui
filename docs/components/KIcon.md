---
title: KIcon
lang: en-US
---

# KIcon

Thanks to unocss capabilities, any Iconify icon can be used in ikun-ui.

Use **any** icons with Pure CSS for ikun-ui.

For more icon information view [@unocss/preset-icons](https://github.com/unocss/unocss/tree/main/packages/preset-icons)

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/icon
```

```bash [yarn]
yarn add @ikun-ui/icon
```

```bash [npm]
npm install @ikun-ui/icon
```

:::

## Icon Collection

Get all icons, try on [icones](https://icones.js.org/)

![](https://raw.githubusercontent.com/antfu/icones/main/screenshots/2.png)
![](https://raw.githubusercontent.com/antfu/icones/main/screenshots/6.png)

## Basic usage

Use the `icon` attribute to define Icon. In addition,the `color` attribute can be used to set the color of the icon.

<demo src="../../../../example/icon/basic.svelte" github='Icon'></demo>

## Icon Props

| Name   | Type      | Default | Description                                               |
| ------ | --------- | ------- | --------------------------------------------------------- |
| icon   | `string`  | `-`     | The class name of the icon, following the unocss standard |
| btn    | `boolean` | `false` | icon as button                                            |
| width  | `string`  | `24px`  | Icon's width                                              |
| height | `string`  | `24px`  | Icon's height                                             |
| color  | `string`  | `-`     | Icon's color                                              |
| cls    | `string`  | `-`     | Additional class                                          |
| attrs  | `any`     | `{}`    | Additional attributes                                     |
