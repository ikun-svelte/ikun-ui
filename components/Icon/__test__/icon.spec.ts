import { tick } from 'svelte';
import { afterEach, expect, test, vi, describe, beforeEach } from 'vitest';
import KIcon from '../src/index.svelte';
import KIconCustomizable from './fixture/customizable.icon.color.test.svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KIcon', () => {
	test('props: icon', async () => {
		const instance = new KIcon({
			target: host,
			props: {
				icon: 'i-carbon-settings'
			}
		});

		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('i-carbon-settings')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: btn', async () => {
		const instance = new KIcon({
			target: host,
			props: {
				icon: 'i-carbon-settings',
				btn: true
			}
		});
		expect(instance).toBeTruthy();
		expect(host.children[0].getAttribute('role') == 'button').toBeTruthy();
		expect(host.innerHTML.includes('k-icon--role-button')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: color', async () => {
		const instance = new KIcon({
			target: host,
			props: {
				icon: 'i-carbon-settings',
				color: 'text-red'
			}
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('text-red')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: width & height', async () => {
		const instance = new KIcon({
			target: host,
			props: {
				icon: 'i-carbon-settings',
				width: '30px',
				height: '30px'
			}
		});
		expect(instance).toBeTruthy();
		expect(host.children[0].style.width).toBe('30px');
		expect(host.children[0].style.height).toBe('30px');
		expect(host.innerHTML).matchSnapshot();
	});

	test('event: should trigger click event', async () => {
		const mockFn = vi.fn();
		let value = false;
		const instance = new KIcon({
			target: host,
			props: {
				icon: 'i-carbon-settings'
			}
		});
		await tick();
		instance.$on('click', (v) => {
			value = v.detail;
			mockFn();
		});
		const input = host.children[0];
		input.click(); // or input.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(instance).toBeTruthy();
		expect(value).toBeTruthy();
		expect(mockFn).toBeCalled();
	});

	test('style: customizable icon color', async () => {
		const instance = new KIconCustomizable({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.children[0].style.background).toBe('red');
		expect(host.innerHTML).matchSnapshot();
	});
});
