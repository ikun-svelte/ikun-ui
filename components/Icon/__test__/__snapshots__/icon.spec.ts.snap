// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KIcon > props: btn 1`] = `"<span class=\\"k-icon--base k-icon--base__dark k-icon--role-button k-icon-transition i-carbon-settings\\" role=\\"button\\" aria-hidden=\\"true\\" style=\\"width: 24px; height: 24px;\\"></span>"`;

exports[`Test: KIcon > props: color 1`] = `"<span class=\\"k-icon--base k-icon-transition i-carbon-settings text-red\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 24px; height: 24px;\\"></span>"`;

exports[`Test: KIcon > props: icon 1`] = `"<span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-settings\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 24px; height: 24px;\\"></span>"`;

exports[`Test: KIcon > props: width & height 1`] = `"<span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-settings\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 30px; height: 30px;\\"></span>"`;

exports[`Test: KIcon > style: customizable icon color 1`] = `"<span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-settings\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"background: red; width: 24px; height: 24px;\\"></span>"`;
